### ISSUE TITLE
Mesos 0.24 based mesos/docker cluster is unstable

### ISSUE THREAD

**Issue by sttts** [mesosphere/kubernetes-mesos#554]
Symptom:
- tasks are scheduler
- the master receives them, sends them to the slave
- in the slave it's logged that they are sent to the executor
- the executor does not see them.

This does not happen all the time. It seems that kube-dns and kube-ui come up fine usually, the same for some more containers. Then suddenly the symptom appears.

**Comment by sttts** [comment]
Even with mesos 0.23 I saw this behaviour now.

Reverting the mesos-go Godeps upgrade in 1455c7b5d7e10fde53f54a3292e2ae538587b958 seems to fix the issue. This is pretty critical also for the new dcos packages if mesos-go is broken.

/cc @jdef

**Comment by jdef** [comment]
How reproducable is this? It would be great to have a detailed log capture (v=3) of the scheduler and executor when this happens.

**Comment by sttts** [comment]
Pretty repducible. Didn't manage to run port forwarding e2e tests
completely ever.

Will add some logs tomorrow.

>

**Comment by sttts** [comment]
The logs:

Scheduler:
```
I1012 13:32:06.589302      50 plugin.go:454] queued pod for scheduling: pfpod
I1012 13:32:06.791060      50 plugin.go:761] Attempting to schedule: &{TypeMeta:{Kind: APIVersion:} ObjectMeta:{Name:pfpod GenerateName: Namespace:e2e-tests-port-forwarding-vhth8 SelfLink:/api/v1/namespaces/e2e-tests-port-forwarding-vhth8/pods/pfpod UID:a19c642f-70e5-11e5-9390-0242c0a841bf ResourceVersion:180 Generation:0 CreationTimestamp:2015-10-12 13:32:06 +0000 UTC DeletionTimestamp:<nil> DeletionGracePeriodSeconds:<nil> Labels:map[name:pfpod] Annotations:map[]} Spec:{Volumes:[{Name:default-token-3vegh VolumeSource:{HostPath:<nil> EmptyDir:<nil> GCEPersistentDisk:<nil> AWSElasticBlockStore:<nil> GitRepo:<nil> Secret:0xc20888de00 NFS:<nil> ISCSI:<nil> Glusterfs:<nil> PersistentVolumeClaim:<nil> RBD:<nil> Cinder:<nil> CephFS:<nil> Flocker:<nil> DownwardAPI:<nil> FC:<nil>}}] Containers:[{Name:portforwardtester Image:gcr.io/google_containers/portforwardtester:1.0 Command:[] Args:[] WorkingDir: Ports:[] Env:[{Name:BIND_PORT Value:80 ValueFrom:<nil>} {Name:EXPECTED_CLIENT_DATA Value:abc ValueFrom:<nil>} {Name:CHUNKS Value:1 ValueFrom:<nil>} {Name:CHUNK_SIZE Value:1 ValueFrom:<nil>} {Name:CHUNK_INTERVAL Value:1 ValueFrom:<nil>}] Resources:{Limits:map[] Requests:map[]} VolumeMounts:[{Name:default-token-3vegh ReadOnly:true MountPath:/var/run/secrets/kubernetes.io/serviceaccount}] LivenessProbe:<nil> ReadinessProbe:<nil> Lifecycle:<nil> TerminationMessagePath:/dev/termination-log ImagePullPolicy:IfNotPresent SecurityContext:<nil> Stdin:false TTY:false}] RestartPolicy:Never TerminationGracePeriodSeconds:0xc20888ddf8 ActiveDeadlineSeconds:<nil> DNSPolicy:ClusterFirst NodeSelector:map[] ServiceAccountName:default NodeName: SecurityContext:0xc20888de30 ImagePullSecrets:[]} Status:{Phase:Pending Conditions:[] Message: Reason: HostIP: PodIP: StartTime:<nil> ContainerStatuses:[]}}
I1012 13:32:06.791503      50 plugin.go:254] Try to schedule pod pfpod
I1012 13:32:06.791764      50 predicate.go:105] trying to match offer with pod e2e-tests-port-forwarding-vhth8/pfpod: cpus: 0.25 mem: 64.00 MB
I1012 13:32:06.791824      50 port_mapping.go:101] Searching for wildcard port in range {8001:21099}
I1012 13:32:06.791899      50 fcfs.go:87] Pod e2e-tests-port-forwarding-vhth8/pfpod accepted offer 20151012-132910-3175196864-5050-1-O37
I1012 13:32:06.791986      50 procurement.go:117] Pod e2e-tests-port-forwarding-vhth8/pfpod without cpu limits is admitted 0.25 cpu shares
I1012 13:32:06.792062      50 procurement.go:120] Pod e2e-tests-port-forwarding-vhth8/pfpod without memory limits is admitted 64.00 MB
I1012 13:32:06.792139      50 procurement.go:132] Recording offer(s) &OfferID{Value:*20151012-132910-3175196864-5050-1-O37,XXX_unrecognized:[],}/e2e-tests-port-forwarding-vhth8 against pod pfpod: cpu: 0.25, mem: 64.00 MB
I1012 13:32:06.792219      50 port_mapping.go:101] Searching for wildcard port in range {8001:21099}
I1012 13:32:06.792299      50 plugin.go:226] prepared pod spec: {TypeMeta:{Kind: APIVersion:} ObjectMeta:{Name:pfpod GenerateName: Namespace:e2e-tests-port-forwarding-vhth8 SelfLink:/api/v1/namespaces/e2e-tests-port-forwarding-vhth8/pods/pfpod UID:a19c642f-70e5-11e5-9390-0242c0a841bf ResourceVersion:180 Generation:0 CreationTimestamp:2015-10-12 13:32:06 +0000 UTC DeletionTimestamp:<nil> DeletionGracePeriodSeconds:<nil> Labels:map[name:pfpod] Annotations:map[k8s.mesosphere.io/taskId:pod.a1bb9819-70e5-11e5-85fd-0242c0a841c2 k8s.mesosphere.io/slaveId:20151012-132910-3175196864-5050-1-S0 k8s.mesosphere.io/offerId:20151012-132910-3175196864-5050-1-O37 k8s.mesosphere.io/executorId:1e136110be445726_k8sm-executor k8s.mesosphere.io/bindingHost:192.168.65.190]} Spec:{Volumes:[{Name:default-token-3vegh VolumeSource:{HostPath:<nil> EmptyDir:<nil> GCEPersistentDisk:<nil> AWSElasticBlockStore:<nil> GitRepo:<nil> Secret:0xc20888de00 NFS:<nil> ISCSI:<nil> Glusterfs:<nil> PersistentVolumeClaim:<nil> RBD:<nil> Cinder:<nil> CephFS:<nil> Flocker:<nil> DownwardAPI:<nil> FC:<nil>}}] Containers:[{Name:portforwardtester Image:gcr.io/google_containers/portforwardtester:1.0 Command:[] Args:[] WorkingDir: Ports:[] Env:[{Name:BIND_PORT Value:80 ValueFrom:<nil>} {Name:EXPECTED_CLIENT_DATA Value:abc ValueFrom:<nil>} {Name:CHUNKS Value:1 ValueFrom:<nil>} {Name:CHUNK_SIZE Value:1 ValueFrom:<nil>} {Name:CHUNK_INTERVAL Value:1 ValueFrom:<nil>}] Resources:{Limits:map[cpu:{Amount:0.250 Format:DecimalSI} memory:{Amount:67108864 Format:BinarySI}] Requests:map[]} VolumeMounts:[{Name:default-token-3vegh ReadOnly:true MountPath:/var/run/secrets/kubernetes.io/serviceaccount}] LivenessProbe:<nil> ReadinessProbe:<nil> Lifecycle:<nil> TerminationMessagePath:/dev/termination-log ImagePullPolicy:IfNotPresent SecurityContext:<nil> Stdin:false TTY:false}] RestartPolicy:Never TerminationGracePeriodSeconds:0xc20888ddf8 ActiveDeadlineSeconds:<nil> DNSPolicy:ClusterFirst NodeSelector:map[] ServiceAccountName:default NodeName: SecurityContext:0xc20888de30 ImagePullSecrets:[]} Status:{Phase:Pending Conditions:[] Message: Reason: HostIP: PodIP: StartTime:<nil> ContainerStatuses:[]}}
I1012 13:32:06.792639      50 plugin.go:180] launching task: "pod.a1bb9819-70e5-11e5-85fd-0242c0a841c2" on target "192.168.65.190" slave "20151012-132910-3175196864-5050-1-S0" for pod "e2e-tests-port-forwarding-vhth8/pfpod", cpu 0.25, mem 64.00 MB
I1012 13:32:06.792851      50 messenger.go:187] Sending message mesos.internal.LaunchTasksMessage to master@192.168.65.189:5050
I1012 13:32:06.793023      50 offers.go:367] Expiring offer 20151012-132910-3175196864-5050-1-O37
I1012 13:32:06.793039      50 offers.go:369] offer will linger: 20151012-132910-3175196864-5050-1-O37
```

Mesos master (0.24):
```
I1012 13:32:06.794621     8 master.hpp:170] Adding task pod.a1bb9819-70e5-11e5-85fd-0242c0a841c2 with resources cpus(*):0.25; mem(*):64 on slave 20151012-132910-3175196864-5050-1-S0 (192.168.65.190)
I1012 13:32:06.794755     8 master.cpp:3069] Launching task pod.a1bb9819-70e5-11e5-85fd-0242c0a841c2 of framework 20151012-132910-3175196864-5050-1-0000 (Kubernetes) at scheduler(1)@192.168.65.194:44819 with resources cpus(*):0.25; mem(*):64 on slave 20151012-132910-3175196864-5050-1-S0 at slave(1)@192.168.65.190:5051 (192.168.65.190)
```

Mesos slave (0.24):
```
I1012 13:32:06.795301   154 slave.cpp:1249] Got assigned task pod.a1bb9819-70e5-11e5-85fd-0242c0a841c2 for framework 20151012-132910-3175196864-5050-1-0000
I1012 13:32:06.795771   154 slave.cpp:1365] Launching task pod.a1bb9819-70e5-11e5-85fd-0242c0a841c2 for framework 20151012-132910-3175196864-5050-1-0000
I1012 13:32:06.796216   154 slave.cpp:1597] Queuing task 'pod.a1bb9819-70e5-11e5-85fd-0242c0a841c2' for executor 1e136110be445726_k8sm-executor of framework '20151012-132910-3175196864-5050-1-0000
I1012 13:32:06.796892   154 cpushare.cpp:392] Updated 'cpu.shares' to 1484 (cpus 1.45) for container de189b1a-9ca1-48fe-a2f2-a0da8fade38d
I1012 13:32:06.797322   155 mem.cpp:355] Updated 'memory.soft_limit_in_bytes' to 498MB for container de189b1a-9ca1-48fe-a2f2-a0da8fade38d
I1012 13:32:06.798092   155 mem.cpp:390] Updated 'memory.limit_in_bytes' to 498MB for container de189b1a-9ca1-48fe-a2f2-a0da8fade38d
I1012 13:32:06.798235   155 slave.cpp:1739] Sending queued task 'pod.a1bb9819-70e5-11e5-85fd-0242c0a841c2' to executor '1e136110be445726_k8sm-executor' of framework 20151012-132910-3175196864-5050-1-0000
I1012 13:32:10.321734   148 slave.cpp:1768] Asked to kill task pod.8426b8fd-70e5-11e5-85fd-0242c0a841c2 of framework 20151012-132910-3175196864-5050-1-0000
```

**Comment by sttts** [comment]
Two go routine dumps of the executor, one where it is broken and one where it still works:  https://gist.github.com/sttts/9f231ffa1546573a3032

**Comment by s-urbaniak** [comment]
@sttts I see no diff at all between the goroutine dumps

**Comment by sttts** [comment]
Yeah, that's the point. I didn't see anything either. I started to dig into
the http server code now. Suddenly no handler is called anymore.

**Comment by sttts** [comment]
/xref https://github.com/mesosphere/kubernetes-mesos/issues/328
/xref https://issues.apache.org/jira/browse/MESOS-2865

What I see pretty much looks like what is described in https://issues.apache.org/jira/browse/MESOS-2865:
- the connection arrives
- net/http starts creating a request
- it reads the first line of the message and block forever.

**Comment by sttts** [comment]
This bug is blocking further work on conformance tests.

**Comment by jdef** [comment]
upstream PR 15735, which fixes this, has been merged

**Comment by jdef** [comment]
fixed by #561
